addons/account/models/account_account.py

```diff
@@ -516,8 +521,13 @@ def _search_new_account_code(self, start_code, cache=None):
         if cache is None:
             cache = {start_code}
 
+        company_ids_to_check = (self.env.company.child_ids | self.env.company.parent_ids).ids
+
         def code_is_available(new_code):
-            return new_code not in cache and not self.search_count([('code', '=', new_code)], limit=1)
+            return (
+                new_code not in cache
+                and not self.sudo().search_count([('company_ids', 'in', company_ids_to_check), ('code', '=', new_code)], limit=1)
```

Comment:
To get the whole hierarchy, use `('company_id', 'child_of', self.env.company.root_id.id)`
Otherwise you might skip some levels of the tree